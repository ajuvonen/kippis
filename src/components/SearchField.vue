<script setup lang="ts">
import {ref} from 'vue';
import {useRouter} from 'vue-router';
import {useCocktailStore} from '@/stores/cocktail';
import {SEARCHABLE_ALCOHOLS} from '@/utils/constants';
import LinkButton from '@/components/LinkButton.vue';
import IconComponent from '@/components/IconComponent.vue';

const props = withDefaults(
  defineProps<{
    omitTitle?: boolean;
    initialValue?: string;
  }>(),
  {
    omitTitle: false,
    initialValue: '',
  },
);

const cocktailStore = useCocktailStore();
const {showRandomCocktail} = cocktailStore;

const router = useRouter();

const searchString = ref(props.initialValue);

const search = () => router.push({name: 'search', query: {searchString: searchString.value}});
</script>

<template>
  <div>
    <label
      for="search-input"
      class="search-field__label"
      :class="{'sr-only': omitTitle}"
      >{{ $t('searchField.label') }}</label
    >
    <div class="search-field__wrapper">
      <input
        id="search-input"
        v-model="searchString"
        type="search"
        maxlength="20"
        class="search-field__input"
        @keypress.enter="search"
      />
      <button class="search-field__search-button" @click="search">
        <IconComponent icon="magnify" />
        <span>{{ $t('searchField.search') }}</span>
      </button>
    </div>
    <ul class="search-field__tag-container" :aria-label="$t('searchField.tagListTitle')">
      <li v-for="{tag} in SEARCHABLE_ALCOHOLS" :key="tag">
        <LinkButton :to="`/search?tag=${tag}`">
          {{ $t(`tags.${tag}`) }}
        </LinkButton>
      </li>
      <li>
        <button @click="showRandomCocktail" data-test-id="search-field__random-button">
          <IconComponent icon="dice" />
          <span>{{ $t('searchField.random') }}</span>
        </button>
      </li>
    </ul>
  </div>
</template>
<style lang="scss" scoped>
.search-field__label {
  @apply block text-center text-3xl my-4;
}

.search-field__wrapper {
  @apply flex justify-center mb-4 divide-x;
}

.search-field__input {
  @apply rounded-tr-none rounded-br-none pb-1 min-w-0;

  &:focus {
    @apply z-10;
  }
}

.search-field__search-button {
  @apply rounded-tl-none rounded-bl-none;
}

.search-field__tag-container {
  @apply flex flex-wrap gap-1 justify-center;
}

@media print {
  .search-field__label,
  .search-field__wrapper,
  .search-field__tag-container {
    @apply hidden;
  }
}
</style>
