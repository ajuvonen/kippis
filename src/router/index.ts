import {createRouter, createWebHistory} from 'vue-router';
import {isEmpty} from 'remeda';
import HomeView from '@/views/HomeView.vue';
import {useCocktailStore} from '@/stores/cocktail';

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes: [
    {
      path: '/',
      name: 'home',
      component: HomeView,
    },
    {
      path: '/search',
      name: 'search',
      props: (route) => ({
        searchString: !route.query.tag ? route.query.searchString : undefined,
        tag: route.query.tag,
      }),
      beforeEnter: (to) => {
        if (isEmpty(to.query)) {
          const cocktailStore = useCocktailStore();
          if (cocktailStore.currentSearch.tag || cocktailStore.currentSearch.searchString) {
            cocktailStore.preventFetch = true;
            return {path: to.path, query: {...cocktailStore.currentSearch}};
          }
        }
      },
      component: () => import('@/views/SearchView.vue'),
    },
    {
      path: '/ingredients',
      name: 'ingredients',
      component: () => import('@/views/IngredientsView.vue'),
    },
  ],
});

export default router;
