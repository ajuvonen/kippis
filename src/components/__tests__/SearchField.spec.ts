import {mount} from '@vue/test-utils';
import {describe, it, expect, vi} from 'vitest';
import router from '@/router';
import SearchField from '@/components/SearchField.vue';

describe('SearchField.vue', () => {
  it('mounts', () => {
    const wrapper = mount(SearchField);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('triggers search when enter key is pressed', async () => {
    const pushSpy = vi.spyOn(router, 'push');
    const wrapper = mount(SearchField);

    const input = wrapper.find('input');
    await input.setValue('test search');
    await input.trigger('keyup.enter');

    expect(pushSpy).toHaveBeenCalledWith({name: 'search', query: {searchString: 'test search'}});
  });

  it('triggers search when search button is clicked', async () => {
    const pushSpy = vi.spyOn(router, 'push');
    const wrapper = mount(SearchField);

    const input = wrapper.find('input');
    await input.setValue('test search');
    await wrapper.find('button').trigger('click');

    expect(pushSpy).toHaveBeenCalledWith({name: 'search', query: {searchString: 'test search'}});
  });

  it('navigates to the correct route when a link button is clicked', async () => {
    const pushSpy = vi.spyOn(router, 'push');
    const wrapper = mount(SearchField);
    const linkButtons = wrapper.findAllComponents({ name: 'LinkButton' });

    for (let i = 0; i < linkButtons.length; i++) {
      const linkButton = linkButtons[i];
      const tag = linkButton.props('to').split('=')[1];
      await linkButton.trigger('click');
      expect(pushSpy).toHaveBeenCalledWith(`/search?tag=${tag}`);
    }
  });
});
