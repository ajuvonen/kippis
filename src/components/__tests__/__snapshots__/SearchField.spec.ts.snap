// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchField.vue > mounts 1`] = `
"<div data-v-b87fbe66=""><label data-v-b87fbe66="" for="search-input" class="search-field__label">Search for Recipes</label>
  <div data-v-b87fbe66="" class="search-field__wrapper"><input data-v-b87fbe66="" id="search-input" type="search" maxlength="20" class="search-field__input"><button data-v-b87fbe66="" class="search-field__search-button"><svg data-v-b87fbe66="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
        <path d="M9.5,3A6.5,6.5 0 0,1 16,9.5C16,11.11 15.41,12.59 14.44,13.73L14.71,14H15.5L20.5,19L19,20.5L14,15.5V14.71L13.73,14.44C12.59,15.41 11.11,16 9.5,16A6.5,6.5 0 0,1 3,9.5A6.5,6.5 0 0,1 9.5,3M9.5,5C7,5 5,7 5,9.5C5,12 7,14 9.5,14C12,14 14,12 14,9.5C14,7 12,5 9.5,5Z" class="transition-all"></path>
      </svg><span data-v-b87fbe66="">Search</span></button></div>
  <ul data-v-b87fbe66="" class="search-field__tag-container" aria-label="Recipe Categories">
    <li data-v-b87fbe66=""><a data-v-c2a26f37="" data-v-b87fbe66="" href="/search?tag=gin" class="link-button">Gin</a></li>
    <li data-v-b87fbe66=""><a data-v-c2a26f37="" data-v-b87fbe66="" href="/search?tag=rum" class="link-button">Rum</a></li>
    <li data-v-b87fbe66=""><a data-v-c2a26f37="" data-v-b87fbe66="" href="/search?tag=vodka" class="link-button">Vodka</a></li>
    <li data-v-b87fbe66=""><a data-v-c2a26f37="" data-v-b87fbe66="" href="/search?tag=whiskey" class="link-button">Whiskey</a></li>
    <li data-v-b87fbe66=""><a data-v-c2a26f37="" data-v-b87fbe66="" href="/search?tag=virgin" class="link-button">Virgin</a></li>
    <li data-v-b87fbe66=""><button data-v-b87fbe66="" data-test-id="search-field__random-button"><svg data-v-b87fbe66="" viewBox="0 0 24 24" aria-hidden="true" class="w-5 h-5">
          <path d="M22,12V20A2,2 0 0,1 20,22H4A2,2 0 0,1 2,20V12A1,1 0 0,1 1,11V8A2,2 0 0,1 3,6H6.17C6.06,5.69 6,5.35 6,5A3,3 0 0,1 9,2C10,2 10.88,2.5 11.43,3.24V3.23L12,4L12.57,3.23V3.24C13.12,2.5 14,2 15,2A3,3 0 0,1 18,5C18,5.35 17.94,5.69 17.83,6H21A2,2 0 0,1 23,8V11A1,1 0 0,1 22,12M4,20H11V12H4V20M20,20V12H13V20H20M9,4A1,1 0 0,0 8,5A1,1 0 0,0 9,6A1,1 0 0,0 10,5A1,1 0 0,0 9,4M15,4A1,1 0 0,0 14,5A1,1 0 0,0 15,6A1,1 0 0,0 16,5A1,1 0 0,0 15,4M3,8V10H11V8H3M13,8V10H21V8H13Z" class="transition-all"></path>
        </svg><span data-v-b87fbe66="">I'm Feeling Lucky</span></button></li>
  </ul>
</div>"
`;
